docs/design/datacontracts/Exception.md

```diff
@@ -1,28 +1,56 @@
-# Contract Thread
+# Contract Exception
 
 This contract is for getting information about exceptions in the process.
 
 ## APIs of contract
 
+```csharp
+record struct ExceptionObjectData(
+    TargetPointer Message,
+    TargetPointer InnerException,
+    TargetPointer StackTrace,
+    TargetPointer WatsonBuckets,
+    TargetPointer StackTraceString,
+    TargetPointer RemoteStackTraceString,
+    int HResult,
+    int XCode);
+```
+
 ``` csharp
 TargetPointer GetExceptionInfo(TargetPointer exception, out TargetPointer nextNestedException);
+ExceptionObjectData GetExceptionObjectData(TargetPointer objectAddress);
 ```
 
 ## Version 1
 
 Data descriptors used:
 - `ExceptionInfo`
+- `ExceptionObject`
```

Comment:
Should we have a consistent naming scheme for data descriptors of managed types? The names of the unmanaged mirrors of managed types are not very consistent.

Here are some examples of managed types that may need data descriptor at some point.
```
System.Exception
System.Runtime.Loader.AssemblyLoadContext
System.Threading.Thread
```

One possible naming scheme is to replace `.` with `_`. It would produce:
`System_Exception`
`System_Runtime_Loader_AssemblyLoadContext`
`System_Threading_Thread`

Reply:
Or omit `.` in the name:
`SystemException`
`SystemRuntimeLoaderAssemblyLoadContext`
`SystemThreadingThread`

Or abbreviate the name space:
`S_Exception`
`SRL_AssemblyLoadContext`
`ST_Thread`

Or drop the namespace completely (I think I like this the most):
`Exception`
`AssemblyLoadContext`
`Thread`

Reply:
> we'd have to differentiate between unmanaged and managed types that have the same name (like Thread) 

The name collisions have been a problem a few times as we are rewriting more of the native runtime in C#. If the name collisions are the only problem, I would say rename the internal types to avoid them. The ultimate picture can be that nearly all types have managed views, and subset of types have unmanaged mirrors.

The name collisions between managed and unmanaged code are also a source of confusion. Trivia question - when you see `ThreadOBJ` in SOS, does it mean managed Thread or unmanaged Thread?